### PR TITLE
StoreGateway: fix store gateway shuffle sharding consistency issue when shard size change by 1

### LINKED ISSUE #5467: Bug in `ring.shuffleShard()` with Zone Aware Replication
### Bug:
There is a bug in how the sub ring for a tenant is obtained in [ring.shuffleShard()](https://github.com/cortexproject/cortex/blob/master/pkg/ring/ring.go#L698). Ideally when shard size increases or decreases by one, the replicas for any block should only change at-most by one instance. But in the cortex implementation more than one instance can change if the `shardSize` is not divisible by `numZones`.

The code for shuffle sharding works as follows [ring.shuffleShard()](https://github.com/cortexproject/cortex/blob/master/pkg/ring/ring.go#L698):
- Get all the zones `r.ringZones`
- Calculate `numInstancesPerZone = int(math.Ceil(float64(shardSize) / float64(numZones)))` 
- For each Zone in `r.ringZones`:
  - Find the first `numInstancesPerZone` unique instances

### Explanation
Assumtions:
- Assume there are 11 store-gateways sg1 to sg11.
- Assume current tenant shard size is 9 and it's increased to 10

#### Initial state
SG_Ring(1)

In the above diagram, the store-gateway tokens are arranged in the order as shown in the first figure. Assume `TENANT_ID` is hashed just before `sg9`'s token.

#### When shardSize=9

ShardSize=9

- When `shardSize = 9`, the shardSize is evenly divisible by number of zones.
```
numInstancesPerZone = int(math.Ceil(float64(shardSize) / float64(numZones)))
numInstancesPerZone = int(math.Ceil(9/3))
numInstancesPerZone = 3
```
- The algorithm for shuffle sharding will pick the first 3 unique instances from each zone with a maximum of 9 instances.
- The subRing for the tenant will be `[sg9, sg5, sg2, sg4, sg1, sg6, sg7, sg8, sg3]`
- We'll have 3 instances per AZ:
  - AZ1: `[sg9, sg2, sg6]`
  - AZ2: `[sg5, sg1, sg7]`
  - AZ3: `[sg4, sg8, sg3]`
- The replicas for the block will be `[sg8, sg9, sg5]`

#### When shardSize=10
ShardSize=10(1)

- When `shardSize=10`, the shardSize is not evenly divisible by number of zones.
```
numInstancesPerZone = int(math.Ceil(float64(shardSize) / float64(numZones)))
numInstancesPerZone = int(math.Ceil(10/3)) = int(math.Ceil(3.33))
numInstancesPerZone = 4
```
- The algorithm for shuffle sharding will pick the first 4 unique instances from each zone with a maximum of 10 instances.
- The subRing for the tenant will be `[sg9, sg5, sg2, sg4, sg1, sg6, sg7, sg8, sg3]`
- Two new instances were added to the subRing: `[sg10, sg11]`
- We'll have 4 instances in AZ3 and AZ2 and only 2 instances in AZ1:
  - AZ1: `[sg9, sg2, sg6, sg11]`
  - AZ2: `[sg15 sg1, sg7, sg10]`
  - AZ2: `[sg4, sg8]`
- The replicas for the block will be `[sg10, sg11, sg8]`
- **The block replicas also changed by 2 when shard size increased from 9 to 11.**

**To Reproduce**
- To reproduce, the store-gateway tenant shard size should be increased by one from a value divisible by number of zones.

**Expected behavior**
- Only one replica should change for any block when tenant shard size increases/decrease by 1.

### Possible Fix
- When the shardSize is not divisible by the zones, only the first zone should get the extra instance.
- The instances per zone shouldn't differ by more than one.

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does**:

Add a new method `ShuffleShardWithZoneStability ` in `ReadRing` interface. It does the same thing as `ShuffleShard` but makes a slight change:
1. It doesn't round up shard size to be multiple of `numZones` when zone awareness is enabled.

If shardSize is divisible by `numZones`, the behavior should be the same.
If shardSize is not divisible by `numZones`, for example, shard size 8 and num zones is 3. Then 2 zones are assigned to have 3 instances and 1 zone has 2 instances. Because we iterate over zones using the same order so we can ensure the consistency.

**Which issue(s) this PR fixes**:
Fixes #5467 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by yeya24** [comment]
`ShuffleShardWithZoneStability` might not be a good name and tbh I am not sure how to name the new shuffle sharding algorithm/change. Any suggestions are welcomed.
For compatibility reason I think we might still want to keep the existing `ShuffleShard` method.

File: pkg/ring/ring.go [hunk 1]
```diff
@@ -778,6 +805,12 @@ func (r *Ring) shuffleShard(identifier string, size int, lookbackPeriod time.Dur
 				break
 			}
 		}
+		if zoneStableSharding && zonesWithExtraInstance > 0 {
```
Comment (wenxu1024) [inline pkg/ring/ring.go:808]: nice change here!

**Review by harry671003 — APPROVED** [review]
Great work! Thanks  for the PR.

**Review by justinjung04 — APPROVED** [review]
Looks good! But I thought this is more of a bug fix in `ring.ShuffleShard` rather than a new feature. I guess you created new function and flag just to avoid any regression for now, but is there a plan to migrate other components to use this func as well?

**Comment by yeya24** [comment]
> But I thought this is more of a bug fix in ring.ShuffleShard rather than a new feature. I guess you created new function and flag just to avoid any regression for now, but is there a plan to migrate other components to use this func as well?

Yes, it is not a new feature but a bug fix specifically for Store Gateway due to consistency check.

I don't know if this is necessary for other components because there is no consistency check to throw 5XXs due to this issue in other components.

No plan for migrations of other components at the moment. We probably need to assess the impact first for each component and see if this is really causing any issue for them. For example, I don't see a big issue of using the existing shuffle sharding algorithm for compactor. We can open another issue to track the assessment and migration.

**Review by alanprot — APPROVED** [review]

**Pull request merged**